pylint/checkers/refactoring.py

```diff
@@ -311,6 +317,54 @@ def leave_functiondef(self, _):
         # new scope = reinitialize the stack of nested blocks
         self._nested_blocks = []
 
+    def visit_raise(self, node):
+        self._check_stop_iteration_inside_generator(node)
+
+    def _check_stop_iteration_inside_generator(self, node):
+        """
+        Check if an exception of type StopIteration is raised inside a generator
+        """
+        scope = node.scope()
+        if not isinstance(scope, astroid.FunctionDef) or not scope.is_generator():
+            return
+        if self._check_raise_inside_tryexcept_catching_stopiter(node):
+            return
+        exception_name = getattr(node.exc, 'name', '')
+        # Get all hand made exceptions defined inside the module
+        # that inherit from StopIteration
+        stopiteration_like_exc = [klass.name for klass in
+                                  node.root().nodes_of_class(astroid.ClassDef)
+                                  if self._check_exception_inherit_from_stopiteration(klass)]
+        if exception_name in ['StopIteration'] + stopiteration_like_exc:
```

Comment:
Hmm I find this dubious. Let's just infer the value of the raised exception, using something as ``safe_infer`` from utils. Then just check that ``_check_exception_inherit_from_stopiteration`` returns True or not on the inferred value.